tools/ci_connector_ops/ci_connector_ops/pipelines/utils.py

```diff
@@ -73,57 +110,78 @@ async def get_file_contents(container: Container, path: str) -> Optional[str]:
         return await container.file(path).contents()
     except QueryError as e:
         if "no such file or directory" not in str(e):
-            # this is the hicky bit of the stopgap because
             # this error could come from a network issue
             raise
     return None
 
 
-# This is a stop-gap solution to capture non 0 exit code on Containers
-# The original issue is tracked here https://github.com/dagger/dagger/issues/3192
+async def get_container_output(container: Container) -> Tuple[str, str]:
+    """Retrieve both stdout and stderr of a container, concurrently.
+
+    Args:
+        container (Container): The container to execute.
+
+    Returns:
+        Tuple[str, str]: The stdout and stderr of the container, respectively.
+
+    Raises:
+        ExecError: If the container exit code is not 0.
+    """
+    async with asyncer.create_task_group() as task_group:
+        soon_stdout = task_group.soonify(container.stdout)
+        soon_stderr = task_group.soonify(container.stderr)
+    return soon_stdout.value, soon_stderr.value
+
+
+async def get_exec_result(container: Container) -> Tuple[int, str, str]:
+    """Retrieve the exit_code along with stdout and stderr of a container by handling the ExecError.
+
+    Note: It is preferrable to not worry about the exit code value and just capture
+    ExecError to handle errors. This is offered as a convenience when the exit code
+    value is actually needed.
+
+    Args:
+        container (Container): The container to execute.
+
+    Returns:
+        Tuple[int, str, str]: The exit_code, stdout and stderr of the container, respectively.
+    """
+    try:
+        return 0, *(await get_container_output(container))
+    except ExecError as e:
+        return e.exit_code, e.stdout, e.stderr
+
+
 async def with_exit_code(container: Container) -> int:
     """Read the container exit code.
 
-    If the exit code is not 0 a QueryError is raised. We extract the non-zero exit code from the QueryError message.
-
     Args:
         container (Container): The container from which you want to read the exit code.
 
     Returns:
         int: The exit code.
     """
     try:
-        await container.exit_code()
-    except QueryError as e:
-        error_message = str(e)
-        if "exit code: " in error_message:
-            exit_code = re.search(r"exit code: (\d+)", error_message)
-            if exit_code:
-                return int(exit_code.group(1))
-            else:
-                return 1
-        raise
+        await container
+    except ExecError as e:
+        return e.exit_code
     return 0
```

Comment:
You can see basic usage in the feature PR:
- https://github.com/dagger/dagger/pull/5184

Note that the lone `raise` still works because this will still raise a `QueryError` on an unexpected network connection failure (for example) since we're only capturing `ExecError` here. It's the equivalent of the `if "exit code: " in error_message` block.